From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Stained Glass OS <ke7oxh@gmail.com>
Date: Thu, 15 Oct 2026 07:19:16 -0700
Subject: [PATCH] rometadata: declare the metadata emit interfaces; E_NOTIMPL
 methods are documented (3318)

DECISIONS-2 NET-META. cor.idl gains IMapToken, IMetaDataEmit, IMetaDataEmit2 and
IMetaDataAssemblyEmit (IIDs and method order checked against the open-source
mingw-w64 cor.h and the interface documentation), for the writer that follows.
IMetaDataDispenserEx::FindAssembly, FindAssemblyModule and OpenScopeOnITypeInfo
are documented by Microsoft as not implemented (E_NOTIMPL), so the FIXMEs become
warnings and the conformance test pins the result.

diff --git a/dlls/rometadata/main.c b/dlls/rometadata/main.c
index 0cabc35..48d8cd9 100644
--- a/dlls/rometadata/main.c
+++ b/dlls/rometadata/main.c
@@ -178,7 +178,8 @@ static HRESULT WINAPI MetaDataDispenser_GetOption(IMetaDataDispenserEx *iface, R
 static HRESULT WINAPI MetaDataDispenser_OpenScopeOnITypeInfo(IMetaDataDispenserEx *iface, ITypeInfo *type_info,
                                                              DWORD open_flags, REFIID riid, IUnknown **obj)
 {
-    FIXME("%p %p %lu %s %p\n", iface, type_info, open_flags, debugstr_guid(riid), obj);
+    /* documented as not implemented: it returns E_NOTIMPL on Windows too */
+    WARN("%p %p %lu %s %p: not implemented\n", iface, type_info, open_flags, debugstr_guid(riid), obj);
     return E_NOTIMPL;
 }
 
@@ -210,8 +211,9 @@ static HRESULT WINAPI MetaDataDispenser_FindAssembly(IMetaDataDispenserEx *iface
                                                      const WCHAR *assembly_name, WCHAR *name, ULONG name_size,
                                                      ULONG *return_length)
 {
-    FIXME("%p %s %s %s %s %p %lu %p\n", iface, debugstr_w(app_base), debugstr_w(private_bin),
-          debugstr_w(global_bin), debugstr_w(assembly_name), name, name_size, return_length);
+    /* documented as not implemented: it returns E_NOTIMPL on Windows too */
+    WARN("%p %s %s %s %s %p %lu %p: not implemented\n", iface, debugstr_w(app_base), debugstr_w(private_bin),
+         debugstr_w(global_bin), debugstr_w(assembly_name), name, name_size, return_length);
     return E_NOTIMPL;
 }
 
@@ -220,8 +222,9 @@ static HRESULT WINAPI MetaDataDispenser_FindAssemblyModule(IMetaDataDispenserEx
                                                            const WCHAR *assembly_name, const WCHAR *module_name,
                                                            WCHAR *name, ULONG name_size, ULONG *return_length)
 {
-    FIXME("%p %s %s %s %s %s %p %lu %p\n", iface, debugstr_w(app_base), debugstr_w(private_bin),
-          debugstr_w(global_bin), debugstr_w(assembly_name), debugstr_w(module_name), name, name_size, return_length);
+    /* documented as not implemented: it returns E_NOTIMPL on Windows too */
+    WARN("%p %s %s %s %s %s %p %lu %p: not implemented\n", iface, debugstr_w(app_base), debugstr_w(private_bin),
+         debugstr_w(global_bin), debugstr_w(assembly_name), debugstr_w(module_name), name, name_size, return_length);
     return E_NOTIMPL;
 }
 
diff --git a/dlls/rometadata/tests/rometadata.c b/dlls/rometadata/tests/rometadata.c
index 85911f8..36f8150 100644
--- a/dlls/rometadata/tests/rometadata.c
+++ b/dlls/rometadata/tests/rometadata.c
@@ -83,6 +83,19 @@ static void test_dispenser_options(void)
     hr = IMetaDataDispenserEx_GetCORSystemDirectory(dispenser, buffer, ARRAY_SIZE(buffer), &length);
     ok(hr == S_OK && length == wcslen(buffer) + 1 && buffer[length - 2] == '\\', "Got unexpected hr %#lx, %s.\n", hr, wine_dbgstr_w(buffer));
 
+    /* these three are documented as not implemented */
+    length = 0;
+    hr = IMetaDataDispenserEx_FindAssembly(dispenser, NULL, NULL, NULL, L"System", buffer, ARRAY_SIZE(buffer), &length);
+    ok(hr == E_NOTIMPL, "Got unexpected hr %#lx.\n", hr);
+    hr = IMetaDataDispenserEx_FindAssemblyModule(dispenser, NULL, NULL, NULL, L"System", L"System.dll", buffer,
+            ARRAY_SIZE(buffer), &length);
+    ok(hr == E_NOTIMPL, "Got unexpected hr %#lx.\n", hr);
+    {
+        IUnknown *unk = (IUnknown *)(ULONG_PTR)1;
+        hr = IMetaDataDispenserEx_OpenScopeOnITypeInfo(dispenser, NULL, 0, &IID_IMetaDataImport, &unk);
+        ok(hr == E_NOTIMPL, "Got unexpected hr %#lx.\n", hr);
+    }
+
     IMetaDataDispenserEx_Release(dispenser);
 }
 
diff --git a/include/cor.idl b/include/cor.idl
index 03b701d..bfc42e9 100644
--- a/include/cor.idl
+++ b/include/cor.idl
@@ -284,6 +284,169 @@ interface IMetaDataAssemblyImport : IUnknown
         [out] IUnknown **ppIUnk, [in] ULONG cMax, [out] ULONG *pcAssemblies);
 }
 
+
+interface IMetaDataAssemblyEmit;
+
+[
+    object,
+    local,
+    uuid(06a3ea8b-0225-11d1-bf72-00c04fc31e12)
+]
+interface IMapToken : IUnknown
+{
+    HRESULT Map([in] ULONG32 tkImp, [in] ULONG32 tkEmit);
+}
+
+typedef struct
+{
+    ULONG32 tkCtor;
+    const void *pCustomAttribute;
+    ULONG cbCustomAttribute;
+} COR_SECATTR;
+
+[
+    object,
+    local,
+    uuid(ba3fee4c-ecb9-4e41-83b7-183fa41cd859)
+]
+interface IMetaDataEmit : IUnknown
+{
+    HRESULT SetModuleProps([in] LPCWSTR szName);
+    HRESULT Save([in] LPCWSTR szFile, [in] DWORD dwSaveFlags);
+    HRESULT SaveToStream([in] IStream *pIStream, [in] DWORD dwSaveFlags);
+    HRESULT GetSaveSize([in] DWORD fSave, [out] DWORD *pdwSaveSize);
+    HRESULT DefineTypeDef([in] LPCWSTR szTypeDef, [in] DWORD dwTypeDefFlags, [in] ULONG32 tkExtends,
+        [in] ULONG32 *rtkImplements, [out] ULONG32 *ptd);
+    HRESULT DefineNestedType([in] LPCWSTR szTypeDef, [in] DWORD dwTypeDefFlags, [in] ULONG32 tkExtends,
+        [in] ULONG32 *rtkImplements, [in] ULONG32 tdEncloser, [out] ULONG32 *ptd);
+    HRESULT SetHandler([in] IUnknown *pUnk);
+    HRESULT DefineMethod([in] ULONG32 td, [in] LPCWSTR szName, [in] DWORD dwMethodFlags, [in] const BYTE *pvSigBlob,
+        [in] ULONG cbSigBlob, [in] ULONG ulCodeRVA, [in] DWORD dwImplFlags, [out] ULONG32 *pmd);
+    HRESULT DefineMethodImpl([in] ULONG32 td, [in] ULONG32 tkBody, [in] ULONG32 tkDecl);
+    HRESULT DefineTypeRefByName([in] ULONG32 tkResolutionScope, [in] LPCWSTR szName, [out] ULONG32 *ptr);
+    HRESULT DefineImportType([in] IMetaDataAssemblyImport *pAssemImport, [in] const void *pbHashValue,
+        [in] ULONG cbHashValue, [in] IMetaDataImport *pImport, [in] ULONG32 tdImport,
+        [in] IMetaDataAssemblyEmit *pAssemEmit, [out] ULONG32 *ptr);
+    HRESULT DefineMemberRef([in] ULONG32 tkImport, [in] LPCWSTR szName, [in] const BYTE *pvSigBlob,
+        [in] ULONG cbSigBlob, [out] ULONG32 *pmr);
+    HRESULT DefineImportMember([in] IMetaDataAssemblyImport *pAssemImport, [in] const void *pbHashValue,
+        [in] ULONG cbHashValue, [in] IMetaDataImport *pImport, [in] ULONG32 mbMember,
+        [in] IMetaDataAssemblyEmit *pAssemEmit, [in] ULONG32 tkParent, [out] ULONG32 *pmr);
+    HRESULT DefineEvent([in] ULONG32 td, [in] LPCWSTR szEvent, [in] DWORD dwEventFlags, [in] ULONG32 tkEventType,
+        [in] ULONG32 mdAddOn, [in] ULONG32 mdRemoveOn, [in] ULONG32 mdFire, [in] ULONG32 *rmdOtherMethods,
+        [out] ULONG32 *pmdEvent);
+    HRESULT SetClassLayout([in] ULONG32 td, [in] DWORD dwPackSize, [in] COR_FIELD_OFFSET *rFieldOffsets,
+        [in] ULONG ulClassSize);
+    HRESULT DeleteClassLayout([in] ULONG32 td);
+    HRESULT SetFieldMarshal([in] ULONG32 tk, [in] const BYTE *pvNativeType, [in] ULONG cbNativeType);
+    HRESULT DeleteFieldMarshal([in] ULONG32 tk);
+    HRESULT DefinePermissionSet([in] ULONG32 tk, [in] DWORD dwAction, [in] const void *pvPermission,
+        [in] ULONG cbPermission, [out] ULONG32 *ppm);
+    HRESULT SetRVA([in] ULONG32 md, [in] ULONG ulRVA);
+    HRESULT GetTokenFromSig([in] const BYTE *pvSig, [in] ULONG cbSig, [out] ULONG32 *pmsig);
+    HRESULT DefineModuleRef([in] LPCWSTR szName, [out] ULONG32 *pmur);
+    HRESULT SetParent([in] ULONG32 mr, [in] ULONG32 tk);
+    HRESULT GetTokenFromTypeSpec([in] const BYTE *pvSig, [in] ULONG cbSig, [out] ULONG32 *ptypespec);
+    HRESULT SaveToMemory([out] void *pbData, [in] ULONG cbData);
+    HRESULT DefineUserString([in] LPCWSTR szString, [in] ULONG cchString, [out] ULONG32 *pstk);
+    HRESULT DeleteToken([in] ULONG32 tkObj);
+    HRESULT SetMethodProps([in] ULONG32 md, [in] DWORD dwMethodFlags, [in] ULONG ulCodeRVA, [in] DWORD dwImplFlags);
+    HRESULT SetTypeDefProps([in] ULONG32 td, [in] DWORD dwTypeDefFlags, [in] ULONG32 tkExtends,
+        [in] ULONG32 *rtkImplements);
+    HRESULT SetEventProps([in] ULONG32 ev, [in] DWORD dwEventFlags, [in] ULONG32 tkEventType, [in] ULONG32 mdAddOn,
+        [in] ULONG32 mdRemoveOn, [in] ULONG32 mdFire, [in] ULONG32 *rmdOtherMethods);
+    HRESULT SetPermissionSetProps([in] ULONG32 tk, [in] DWORD dwAction, [in] const void *pvPermission,
+        [in] ULONG cbPermission, [out] ULONG32 *ppm);
+    HRESULT DefinePinvokeMap([in] ULONG32 tk, [in] DWORD dwMappingFlags, [in] LPCWSTR szImportName,
+        [in] ULONG32 mrImportDLL);
+    HRESULT SetPinvokeMap([in] ULONG32 tk, [in] DWORD dwMappingFlags, [in] LPCWSTR szImportName,
+        [in] ULONG32 mrImportDLL);
+    HRESULT DeletePinvokeMap([in] ULONG32 tk);
+    HRESULT DefineCustomAttribute([in] ULONG32 tkOwner, [in] ULONG32 tkCtor, [in] const void *pCustomAttribute,
+        [in] ULONG cbCustomAttribute, [out] ULONG32 *pcv);
+    HRESULT SetCustomAttributeValue([in] ULONG32 pcv, [in] const void *pCustomAttribute,
+        [in] ULONG cbCustomAttribute);
+    HRESULT DefineField([in] ULONG32 td, [in] LPCWSTR szName, [in] DWORD dwFieldFlags, [in] const BYTE *pvSigBlob,
+        [in] ULONG cbSigBlob, [in] DWORD dwCPlusTypeFlag, [in] const void *pValue, [in] ULONG cchValue,
+        [out] ULONG32 *pmd);
+    HRESULT DefineProperty([in] ULONG32 td, [in] LPCWSTR szProperty, [in] DWORD dwPropFlags, [in] const BYTE *pvSig,
+        [in] ULONG cbSig, [in] DWORD dwCPlusTypeFlag, [in] const void *pValue, [in] ULONG cchValue,
+        [in] ULONG32 mdSetter, [in] ULONG32 mdGetter, [in] ULONG32 *rmdOtherMethods, [out] ULONG32 *pmdProp);
+    HRESULT DefineParam([in] ULONG32 md, [in] ULONG ulParamSeq, [in] LPCWSTR szName, [in] DWORD dwParamFlags,
+        [in] DWORD dwCPlusTypeFlag, [in] const void *pValue, [in] ULONG cchValue, [out] ULONG32 *ppd);
+    HRESULT SetFieldProps([in] ULONG32 fd, [in] DWORD dwFieldFlags, [in] DWORD dwCPlusTypeFlag,
+        [in] const void *pValue, [in] ULONG cchValue);
+    HRESULT SetPropertyProps([in] ULONG32 pr, [in] DWORD dwPropFlags, [in] DWORD dwCPlusTypeFlag,
+        [in] const void *pValue, [in] ULONG cchValue, [in] ULONG32 mdSetter, [in] ULONG32 mdGetter,
+        [in] ULONG32 *rmdOtherMethods);
+    HRESULT SetParamProps([in] ULONG32 pd, [in] LPCWSTR szName, [in] DWORD dwParamFlags, [in] DWORD dwCPlusTypeFlag,
+        [in] const void *pValue, [in] ULONG cchValue);
+    HRESULT DefineSecurityAttributeSet([in] ULONG32 tkObj, [in] COR_SECATTR *rSecAttrs, [in] ULONG cSecAttrs,
+        [out] ULONG *pulErrorAttr);
+    HRESULT ApplyEditAndContinue([in] IUnknown *pImport);
+    HRESULT TranslateSigWithScope([in] IMetaDataAssemblyImport *pAssemImport, [in] const void *pbHashValue,
+        [in] ULONG cbHashValue, [in] IMetaDataImport *pImport, [in] const BYTE *pbSigBlob, [in] ULONG cbSigBlob,
+        [in] IMetaDataAssemblyEmit *pAssemEmit, [in] IMetaDataEmit *emit, [out] BYTE *pvTranslatedSig,
+        [in] ULONG cbTranslatedSigMax, [out] ULONG *pcbTranslatedSig);
+    HRESULT SetMethodImplFlags([in] ULONG32 md, [in] DWORD dwImplFlags);
+    HRESULT SetFieldRVA([in] ULONG32 fd, [in] ULONG ulRVA);
+    HRESULT Merge([in] IMetaDataImport *pImport, [in] IMapToken *pHostMapToken, [in] IUnknown *pHandler);
+    HRESULT MergeEnd();
+}
+
+[
+    object,
+    local,
+    uuid(f5dd9950-f693-42e6-830e-7b833e8146a9)
+]
+interface IMetaDataEmit2 : IMetaDataEmit
+{
+    HRESULT DefineMethodSpec([in] ULONG32 tkParent, [in] const BYTE *pvSigBlob, [in] ULONG cbSigBlob,
+        [out] ULONG32 *pmi);
+    HRESULT GetDeltaSaveSize([in] DWORD fSave, [out] DWORD *pdwSaveSize);
+    HRESULT SaveDelta([in] LPCWSTR szFile, [in] DWORD dwSaveFlags);
+    HRESULT SaveDeltaToStream([in] IStream *pIStream, [in] DWORD dwSaveFlags);
+    HRESULT SaveDeltaToMemory([out] void *pbData, [in] ULONG cbData);
+    HRESULT DefineGenericParam([in] ULONG32 tk, [in] ULONG ulParamSeq, [in] DWORD dwParamFlags, [in] LPCWSTR szname,
+        [in] DWORD reserved, [in] ULONG32 *rtkConstraints, [out] ULONG32 *pgp);
+    HRESULT SetGenericParamProps([in] ULONG32 gp, [in] DWORD dwParamFlags, [in] LPCWSTR szName, [in] DWORD reserved,
+        [in] ULONG32 *rtkConstraints);
+    HRESULT ResetENCLog();
+}
+
+[
+    object,
+    local,
+    uuid(211ef15b-5317-4438-b196-dec87b887693)
+]
+interface IMetaDataAssemblyEmit : IUnknown
+{
+    HRESULT DefineAssembly([in] const void *pbPublicKey, [in] ULONG cbPublicKey, [in] ULONG ulHashAlgId,
+        [in] LPCWSTR szName, [in] const ASSEMBLYMETADATA *pMetaData, [in] DWORD dwAssemblyFlags,
+        [out] ULONG32 *pma);
+    HRESULT DefineAssemblyRef([in] const void *pbPublicKeyOrToken, [in] ULONG cbPublicKeyOrToken,
+        [in] LPCWSTR szName, [in] const ASSEMBLYMETADATA *pMetaData, [in] const void *pbHashValue,
+        [in] ULONG cbHashValue, [in] DWORD dwAssemblyRefFlags, [out] ULONG32 *pmdar);
+    HRESULT DefineFile([in] LPCWSTR szName, [in] const void *pbHashValue, [in] ULONG cbHashValue,
+        [in] DWORD dwFileFlags, [out] ULONG32 *pmdf);
+    HRESULT DefineExportedType([in] LPCWSTR szName, [in] ULONG32 tkImplementation, [in] ULONG32 tkTypeDef,
+        [in] DWORD dwExportedTypeFlags, [out] ULONG32 *pmdct);
+    HRESULT DefineManifestResource([in] LPCWSTR szName, [in] ULONG32 tkImplementation, [in] DWORD dwOffset,
+        [in] DWORD dwResourceFlags, [out] ULONG32 *pmdmr);
+    HRESULT SetAssemblyProps([in] ULONG32 pma, [in] const void *pbPublicKey, [in] ULONG cbPublicKey,
+        [in] ULONG ulHashAlgId, [in] LPCWSTR szName, [in] const ASSEMBLYMETADATA *pMetaData,
+        [in] DWORD dwAssemblyFlags);
+    HRESULT SetAssemblyRefProps([in] ULONG32 ar, [in] const void *pbPublicKeyOrToken,
+        [in] ULONG cbPublicKeyOrToken, [in] LPCWSTR szName, [in] const ASSEMBLYMETADATA *pMetaData,
+        [in] const void *pbHashValue, [in] ULONG cbHashValue, [in] DWORD dwAssemblyRefFlags);
+    HRESULT SetFileProps([in] ULONG32 file, [in] const void *pbHashValue, [in] ULONG cbHashValue,
+        [in] DWORD dwFileFlags);
+    HRESULT SetExportedTypeProps([in] ULONG32 ct, [in] ULONG32 tkImplementation, [in] ULONG32 tkTypeDef,
+        [in] DWORD dwExportedTypeFlags);
+    HRESULT SetManifestResourceProps([in] ULONG32 mr, [in] ULONG32 tkImplementation, [in] DWORD dwOffset,
+        [in] DWORD dwResourceFlags);
+}
+
 cpp_quote("BOOL        WINAPI _CorDllMain(HINSTANCE,DWORD,LPVOID);")
 cpp_quote("__int32     WINAPI _CorExeMain(void);")
 cpp_quote("__int32     WINAPI _CorExeMain2(PBYTE,DWORD,LPWSTR,LPWSTR,LPWSTR);")
